collect_app/src/main/java/org/odk/collect/android/formmanagement/FormsUpdater.kt

```diff
@@ -0,0 +1,182 @@
+    private val formsRepositoryProvider: FormsRepositoryProvider,
+    private val formSourceProvider: FormSourceProvider,
+    private val syncStatusAppState: SyncStatusAppState,
+    private val instancesRepositoryProvider: InstancesRepositoryProvider,
+    private val changeLockProvider: ChangeLockProvider
+) {
+
+    /**
+     * Downloads updates for the project's already downloaded forms. If Automatic download is
+     * disabled the user will just be notified that there are updates available.
+     */
+    fun downloadUpdates(projectId: String) {
+        val sandbox = getProjectSandbox(projectId)
+
+        val diskFormsSynchronizer = diskFormsSynchronizer(sandbox)
+        val serverFormsDetailsFetcher = serverFormsDetailsFetcher(sandbox, diskFormsSynchronizer)
+        val formDownloader = formDownloader(sandbox, analytics)
+
+        try {
+            val serverForms: List<ServerFormDetails> = serverFormsDetailsFetcher.fetchFormDetails()
+            val updatedForms =
+                serverForms.stream().filter { obj: ServerFormDetails -> obj.isUpdated }
+                    .collect(Collectors.toList())
+            if (updatedForms.isNotEmpty()) {
+                if (sandbox.generalSettings.getBoolean(GeneralKeys.KEY_AUTOMATIC_UPDATE)) {
+                    val formUpdateDownloader = FormUpdateDownloader()
+                    val results = formUpdateDownloader.downloadUpdates(
+                        updatedForms,
+                        sandbox.formsLock,
+                        formDownloader,
+                        TranslationHandler.getString(context, R.string.success),
+                        TranslationHandler.getString(context, R.string.failure)
+                    )
+
+                    notifier.onUpdatesDownloaded(results)
+                } else {
+                    notifier.onUpdatesAvailable(updatedForms)
+                }
+            }
+
+            context.contentResolver.notifyChange(FormsProviderAPI.CONTENT_URI, null)
+        } catch (_: FormSourceException) {
+            // Ignored
+        }
+    }
+
+    /**
+     * Downloads new forms, updates existing forms and deletes forms that are no longer part of
+     * the project's form list.
+     */
+    fun matchFormsWithServer(projectId: String): Boolean {
+        val sandbox = getProjectSandbox(projectId)
+
+        val diskFormsSynchronizer = diskFormsSynchronizer(sandbox)
+        val serverFormsDetailsFetcher = serverFormsDetailsFetcher(sandbox, diskFormsSynchronizer)
+        val formDownloader = formDownloader(sandbox, analytics)
+
+        val serverFormsSynchronizer = ServerFormsSynchronizer(
+            serverFormsDetailsFetcher,
+            sandbox.formsRepository,
+            sandbox.instancesRepository,
+            formDownloader
+        )
+
+        return sandbox.formsLock.withLock { acquiredLock ->
+            if (acquiredLock) {
+                syncStatusAppState.startSync()
```

Comment:
I realized that we need to rework this so that the "syncing" state works on a per-project basis. I've added this as work to do post beta. The worst case right now is that someone won't be able to manually refresh their project while another one is syncing in the background - fine for beta, but we need to fix before launch.